azure-iot-device/azure/iot/device/iothub/pipeline/constant.py

```diff
@@ -10,3 +10,5 @@
 C2D_MSG = "c2d"
 INPUT_MSG = "input"
 METHODS = "methods"
+TWIN = "twin"
+TWIN_PATCHES = "twin_patches"
```

Comment:
Not sure if we actually need these to be separate

Reply:
This is somewhat internal and it needs to be separate because they represent two subsets of functionality.  TWIN is for the app-initiated "GetTwin" and 'SendTwinReportedPropertiesPatch".  TWIN_PATCHES Is for the service-initiated ReceiveTwinDesiredPropertiesPatch.  The app-initiated ops need the response topic and the service-initiated op needs a downstream topic.  

(This is also related to something that Tim told you -- by separating these, the user can use the twin for desired properties, or they can use it for reported properties, but they're not required to subscribe for both.)

---
In reply to: [288334435](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288334435) [](ancestors = 288334435)

Reply:
My understanding of what Tim said was the opposite - that Twin always has two AMQP links going.

But even if I'm mistaken, and we do need separate ones, SEND_TWIN and RECEIVE_TWIN might be better, since neither TWIN nor TWIN_PATCHES clearly indicates a direction (patches could be sent or received for instance).

Furthermore, why do we need a TWIN/SEND_TWIN feature enable/disable? We don't need to enable sending telemetry, or sending a method response. Why do we need to enable the sending of a twin patch, or the request for a twin?

---
In reply to: [288657819](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288657819) [](ancestors = 288657819,288334435)

Reply:
I think (if I heard correctly) that Tim way saying that the Java implementation had both links going, but it was theoretically possible to only use one.

This whole mess happens because MQTT and AMQP don't support the idea of replying to something.  Sending a request is one operation (upstream) and receiving the response means listening for a second operation (downstream).  We don't need to enable anything to _send_ the patch or to send the GET request.  We need the enable because we need to _listen_ for the response.  For GET, the request gets published on one topic,then we have to listen on a different topic for the response.   The "TWIN" feature name enables us to listen for the response., and it just happens that GET_TWIN and SEND_TWIN_PATCH both use the same downstream channel.  

I really struggled with this, because it's not really SEND_TWIN vs RECEIVE_TWIN, its GET_AND_SEND_TWIN vs RECEIVE_TWIN.  There's a difference, but it's hard to explain and hard to understand.  Maybe there should be one thing to enable, and it enables both scenarios under the covers.  Or maybe I make it 3 features and 2 of them happen to overlap.  That would be a little simpler (because there's a 1:1 mapping between feature and topic).



---
In reply to: [288770383](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288770383) [](ancestors = 288770383,288657819,288334435)

Reply:
I followed up with Tim (and also chatted with Zoltan) and it does seem like the operations are not independent. But I think we should check with service.

---
In reply to: [288776071](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288776071) [](ancestors = 288776071,288770383,288657819,288334435)

Reply:
sure, but don't forget that this feature_name stuff is all internal.  We're not exposing any of it to users.  

---
In reply to: [288790187](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288790187) [](ancestors = 288790187,288776071,288770383,288657819,288334435)